### PR TITLE
Add mock test for VerifyUpgradeAndUpdateState

### PR DESCRIPTION
**Pull request by DimitrisJim**
## Description

closes: #4832 

### Commit Message / Changelog Entry

```text
type: commit message
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/light-clients/08-wasm/types/upgrade_test.go [hunk 1]
```diff
@@ -98,536 +106,145 @@ func (suite *TypesTestSuite) TestVerifyUpgradeGrandpa() {
...
+				// set upgraded client state to solomachine client state
+				upgradedClient = &solomachine.ClientState{}
+			},
+			func(err error) {
+				suite.Require().Error(err)
+				suite.Require().ErrorIs(err, clienttypes.ErrInvalidClient)
 
-// 			newChainID, err = clienttypes.SetRevisionNumber(tmClientState.ChainId, revisionNumber+1)
-// 			suite.Require().NoError(err)
+				// verify client state is unchanged
```
Comment (crodriguezvega) [inline modules/light-clients/08-wasm/types/upgrade_test.go:180]: Should we also verify that consensus state is unchanged?

File: modules/light-clients/08-wasm/types/upgrade_test.go [hunk 2]
```diff
@@ -98,536 +106,145 @@ func (suite *TypesTestSuite) TestVerifyUpgradeGrandpa() {
...
+			func(err error) {
+				suite.Require().Error(err)
+				suite.Require().ErrorIs(err, clienttypes.ErrInvalidConsensus)
 
-// 			var tmUpgradedClient exported.ClientState
-// 			tmUpgradedClient = ibctm.NewClientState(newChainID, ibctm.DefaultTrustLevel, trustingPeriod, ubdPeriod+trustingPeriod, maxClockDrift, clienttypes.NewHeight(revisionNumber+1, clientState.GetLatestHeight().GetRevisionHeight()+1), commitmenttypes.GetSDKSpecs(), upgradePath)
-// 			tmUpgradedClient = tmUpgradedClient.ZeroCustomFields()
-// 			tmUpgradedClientBz, err := clienttypes.MarshalClientState(suite.chainA.App.AppCodec(), tmUpgradedClient)
-// 			suite.Require().NoError(err)
+				// verify client state is unchanged
```
Comment (crodriguezvega) [inline modules/light-clients/08-wasm/types/upgrade_test.go:195]: ditto.

File: modules/light-clients/08-wasm/types/upgrade_test.go [hunk 3]
```diff
@@ -98,536 +106,145 @@ func (suite *TypesTestSuite) TestVerifyUpgradeGrandpa() {
...
+				})
+			},
+			func(err error) {
+				suite.Require().Error(err)
+				suite.Require().ErrorIs(err, contractError)
 
-// 			upgradedClient = types.NewClientState(tmUpgradedClientBz, clientState.CodeHash, clienttypes.NewHeight(revisionNumber+1, clientState.GetLatestHeight().GetRevisionHeight()+1))
-// 			upgradedClientBz, err = clienttypes.MarshalClientState(suite.chainA.App.AppCodec(), upgradedClient)
-// 			suite.Require().NoError(err)
+				// verify client state is unchanged
```
Comment (crodriguezvega) [inline modules/light-clients/08-wasm/types/upgrade_test.go:211]: Also here. :)

**Review by crodriguezvega — APPROVED** [review]
Great, @DimitrisJim! Just one more comment: should we add a test case for [this](https://github.com/cosmos/ibc-go/blob/8a0eda1816bec109b17db08e7fc13d7fc41c08ba/modules/light-clients/08-wasm/types/upgrade.go#L41-L44)?

**Comment by DimitrisJim** [comment]
> should we add a test case for [this](https://github.com/cosmos/ibc-go/blob/8a0eda1816bec109b17db08e7fc13d7fc41c08ba/modules/light-clients/08-wasm/types/upgrade.go#L41-L44)?

Ah, yea I actually did think of this (my cov OCD hates it) but its actually pretty tricky to get Marshal to fail. Issue is you get a typed value passed in (i.e `InstantiateMessage`) which I can't think of a way to populate in order to have marshalling fail (You typically need either a cyclic reference or some type [that isn't in the supported types](https://github.com/golang/go/blob/b5f87b5407916c4049a3158cc944cebfd7a883a9/src/encoding/json/encode.go#L396), like channels).

Open to hacky suggestions if people have em.

**Comment by DimitrisJim** [comment]
> Just one more comment: should we add a test case for [this](https://github.com/cosmos/ibc-go/blob/8a0eda1816bec109b17db08e7fc13d7fc41c08ba/modules/light-clients/08-wasm/types/upgrade.go#L41-L44)?

We'll be removing that check in #4368 (removing height-check and moving it to 02-client). Open to adding it until then if people think its necessary.

**Comment by crodriguezvega** [comment]
> > Just one more comment: should we add a test case for [this](https://github.com/cosmos/ibc-go/blob/8a0eda1816bec109b17db08e7fc13d7fc41c08ba/modules/light-clients/08-wasm/types/upgrade.go#L41-L44)?
> 
> We'll be removing that check in #4368 (removing height-check and moving it to 02-client). Open to adding it until then if people think its necessary.

Oh, yes, forgot about that. Happy to skip the test then.

File: modules/light-clients/08-wasm/types/upgrade_test.go (hunk 1 above)
Comment (DimitrisJim) [reply modules/light-clients/08-wasm/types/upgrade_test.go]: Done! Made move back to the `expErr` workflow, `expResult` was just more unnecessary repetition.

File: modules/light-clients/08-wasm/types/upgrade_test.go [hunk 4]
```diff
@@ -98,536 +106,137 @@ func (suite *TypesTestSuite) TestVerifyUpgradeGrandpa() {
...
-// 			upgradedConsStateBz, err = clienttypes.MarshalConsensusState(suite.chainA.App.AppCodec(), upgradedConsState)
-// 			suite.Require().NoError(err)
+			proofUpgradedClient = []byte("proof upgraded client")
+			proofUpgradedConsState = []byte("proof upgraded consensus state")
 
-// 			tc.setup()
+			err = clientState.VerifyUpgradeAndUpdateState(
+				suite.ctx,
+				suite.chainA.Codec,
+				suite.store,
```
Comment (colin-axner) [inline modules/light-clients/08-wasm/types/upgrade_test.go:209]: should we be referencing the updated context/store here? Or it it the intention to use the static values set by setup?

File: modules/light-clients/08-wasm/types/upgrade_test.go [hunk 5]
```diff
@@ -98,536 +106,137 @@ func (suite *TypesTestSuite) TestVerifyUpgradeGrandpa() {
...
-// 			},
-// 			expPass: false,
-// 		},
-// 		{
-// 			name: "unsuccessful upgrade: consensus state for upgrade height cannot be found",
-// 			setup: func() {
-// 				// upgrade Height is at next block
-// 				lastHeight = clienttypes.NewHeight(0, uint64(suite.chainB.GetContext().BlockHeight()+100))
+func (suite *TypesTestSuite) TestVerifyUpgradeAndUpdateState() {
+	// TODO: PR 4934 exposes a ErrMockContract that can be re-used
```
Comment (charleenfei) [inline modules/light-clients/08-wasm/types/upgrade_test.go:110]: ❤️

File: modules/light-clients/08-wasm/types/upgrade_test.go [hunk 6]
```diff
@@ -98,536 +106,137 @@ func (suite *TypesTestSuite) TestVerifyUpgradeGrandpa() {
...
+				suite.Require().Equal(upgradedConsState, clienttypes.MustUnmarshalConsensusState(suite.chainA.Codec, consStateBz))
+			} else {
+				suite.Require().Error(err)
+				suite.Require().ErrorIs(err, tc.expErr)
 
-// 				clientState := suite.chainA.GetClientState(path.EndpointA.ClientID)
-// 				suite.Require().NotNil(clientState, "verify upgrade failed on valid case: %s", tc.name)
+				// verify client state is unchanged
+				clientState := clienttypes.MustUnmarshalClientState(suite.chainA.Codec, suite.store.Get(host.ClientStateKey()))
+				suite.Require().Equal(clientState, endpoint.GetClientState().(*types.ClientState))
```
Comment (colin-axner) [inline modules/light-clients/08-wasm/types/upgrade_test.go:234]: I think this is circular logic, `endpoint.GetClientState()` gets the client from the store. We would need to reference a client state var grabbed from state before the verify upgrade call

File: modules/light-clients/08-wasm/types/upgrade_test.go [hunk 7]
```diff
@@ -98,536 +106,137 @@ func (suite *TypesTestSuite) TestVerifyUpgradeGrandpa() {
...
-// 				suite.Require().True(found)
-// 			} else {
-// 				suite.Require().Error(err, "verify upgrade passed on invalid case: %s", tc.name)
-// 			}
-// 		})
-// 	}
-// }
+				// verify consensus state is unchanged
+				consensusState := clienttypes.MustUnmarshalConsensusState(suite.chainA.Codec, suite.store.Get(host.ConsensusStateKey(clientState.GetLatestHeight())))
+				suite.Require().Equal(consensusState, endpoint.GetConsensusState(clientState.GetLatestHeight()).(*types.ConsensusState))
```
Comment (colin-axner) [inline modules/light-clients/08-wasm/types/upgrade_test.go:238]: ditto

File: modules/light-clients/08-wasm/types/upgrade_test.go (hunk 6 above)
Comment (colin-axner) [reply modules/light-clients/08-wasm/types/upgrade_test.go:234]: regardless, I think we can remove this bit of code. An error returned to baseapp reverts all state changes anyways. We only need to check that state isn't changed on failure when we are using cached contexts and returning successful transaction to baseapp despite some failure route in the execution trace

**Review by colin-axner — APPROVED** [review]
LGTM, but see comments on state checks on error cases

File: modules/light-clients/08-wasm/types/upgrade_test.go (hunk 4 above)
Comment (DimitrisJim) [reply modules/light-clients/08-wasm/types/upgrade_test.go:209]: Intention is to use static values, other tests go the same route. Pretty sure this should play no role in the mocking case, correct? All we basically test for here is that the store we pass in eventually does get written to.

**Pull request merged**